### PR TITLE
Softmax implementation

### PR DESCRIPTION
**Pull request by sreenikSS**

### CODE REVIEW

File: src/mlpack/methods/ann/model_parser.hpp [hunk 1]
```diff
@@ -0,0 +1,294 @@
+/**
+ * @file model_parser.hpp
```
Comment (zoq) [inline src/mlpack/methods/ann/model_parser.hpp]: Should this be part of the PR?

Comment (sreenikSS) [reply src/mlpack/methods/ann/model_parser.hpp]: No this shouldn't have been. I'm a little new to branching, the old changes also got reflected in the new branch somehow. Got to figure that out.

Comment (zoq) [reply src/mlpack/methods/ann/model_parser.hpp]: I see, what I usually do, is to branch from master. But that also means I don't do any work in the master branch.

Comment (zoq) [reply src/mlpack/methods/ann/model_parser.hpp]: Looks like in your case you already did some changes in the master branch, so maybe it makes sense to first move those changes into another branch and remove the commits from the master branch.

**Comment by zoq** [comment]
Hello @sreenikSS any update on this one, do you need any help to move the model parser code out of the PR?

**Comment by codeboy5** [comment]
@zoq Can I work on this issue, it seems abandoned to me.

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

**Comment by kartikdutt18** [comment]
Hey there, This code touches the loss function portion of mlpack's codebase. Recently #2339 was merged where the switch was made to templated return type instead of just using double. So before we merge this, kindly make the changes for the same. I'm pasting this in all relevant PRs. Thanks a lot 👍

**Comment by mrityunjay-tripathi** [comment]
Hey @birm, as #2351 is merged already, I think we can close this PR now.

**Comment by birm** [comment]
Good catch! I guess github's closing keywords only work on issues, not other prs.
